File: cwe_comparative/CWE-787_out_of_bounds_write/vulnerable.go
// CWE-787: Out-of-bounds Write
// Go version - runtime panic on out-of-bounds

package main

import "fmt"

func vulnerableCopy(input []byte) {
	buffer := make([]byte, 10)
	// Go will panic at runtime if input > 10
	copy(buffer, input) // Actually safe - copy truncates

	// This would panic:
	for i, b := range input {
		buffer[i] = b // PANIC if i >= 10
	}
	fmt.Printf("Copied: %s\n", buffer)
}

func main() {
	vulnerableCopy([]byte("This is too long!"))
}

File: cwe_comparative/CWE-787_out_of_bounds_write/vulnerable.rs
// CWE-787: Out-of-bounds Write
// Rust version - compile-time prevention in safe Rust

fn vulnerable_copy(input: &[u8]) {
    let mut buffer = [0u8; 10];

    // This would be a compile error - can't just index past bounds
    // buffer[..input.len()].copy_from_slice(input); // ERROR if input.len() > 10

    // Safe version - Rust forces you to handle the bounds:
    let len = std::cmp::min(input.len(), buffer.len());
    buffer[..len].copy_from_slice(&input[..len]);

    println!("Copied: {:?}", &buffer[..len]);
}

// With unsafe, you CAN have buffer overflows:
unsafe fn unsafe_copy(input: &[u8]) {
    let mut buffer = [0u8; 10];
    // This compiles but is UB if input.len() > 10
    std::ptr::copy_nonoverlapping(input.as_ptr(), buffer.as_mut_ptr(), input.len());
}

fn main() {
    vulnerable_copy(b"This is too long for the buffer!");
}
